models/git/branch.go

```diff
@@ -381,24 +391,100 @@ func RenameBranch(ctx context.Context, repo *repo_model.Repository, from, to str
 	return committer.Commit()
 }
 
+type FindRecentlyPushedNewBranchesOptions struct {
+	db.ListOptions
+	Actor           *user_model.User
+	Repo            *repo_model.Repository
+	BaseRepo        *repo_model.Repository
+	CommitAfterUnix int64
+}
+
+type RecentlyPushedNewBranch struct {
+	BranchName       string
+	BranchCompareURL string
+	CommitTime       timeutil.TimeStamp
+}
+
 // FindRecentlyPushedNewBranches return at most 2 new branches pushed by the user in 6 hours which has no opened PRs created
-// except the indicate branch
-func FindRecentlyPushedNewBranches(ctx context.Context, repoID, userID int64, excludeBranchName string) (BranchList, error) {
-	branches := make(BranchList, 0, 2)
-	subQuery := builder.Select("head_branch").From("pull_request").
-		InnerJoin("issue", "issue.id = pull_request.issue_id").
-		Where(builder.Eq{
-			"pull_request.head_repo_id": repoID,
-			"issue.is_closed":           false,
+// opts.Actor should not be nil
+// if opts.CommitAfterUnix is 0, we will find the branches committed in recently 6 hours
+// if opts.ListOptions is not set, we will only display top 2 latest branch
+func FindRecentlyPushedNewBranches(ctx context.Context, opts *FindRecentlyPushedNewBranchesOptions) ([]*RecentlyPushedNewBranch, error) {
+	// find all related repo ids
+	repoOpts := repo_model.SearchRepoOptions{
+		Actor:      opts.Actor,
+		Private:    true,
+		AllPublic:  false, // Include also all public repositories of users and public organisations
+		AllLimited: false, // Include also all public repositories of limited organisations
+		Fork:       util.OptionalBoolTrue,
+		ForkFrom:   opts.BaseRepo.ID,
+		Archived:   util.OptionalBoolFalse,
+	}
+	repoCond := repo_model.SearchRepositoryCondition(&repoOpts).And(repo_model.AccessibleRepositoryCondition(opts.Actor, unit.TypeCode))
+	if opts.Repo == opts.BaseRepo {
+		// should also include the base repo's branches
+		repoCond = repoCond.Or(builder.Eq{"id": opts.BaseRepo.ID})
+	} else {
+		// in fork repo, we only detect the fork repo's branch
+		repoCond = repoCond.And(builder.Eq{"id": opts.Repo.ID})
+	}
+	repoIDs := builder.Select("id").From("repository").Where(repoCond)
+
+	// find branches which have already created PRs
+	prBranchIds := builder.Select("branch.id").From("branch").
+		InnerJoin("pull_request", "branch.name = pull_request.head_branch AND branch.repo_id = pull_request.head_repo_id").
+		Where(builder.And(
+			builder.Eq{"pull_request.base_repo_id": opts.BaseRepo.ID},
+			builder.Eq{"pull_request.base_branch": opts.BaseRepo.DefaultBranch},
+			builder.In("pull_request.head_repo_id", repoIDs),
+		))
+
+	if opts.CommitAfterUnix == 0 {
+		opts.CommitAfterUnix = time.Now().Add(-time.Hour * 6).Unix()
```

Comment:
```suggestion
		opts.CommitAfterUnix = time.Now().Add(-time.Hour * 2).Unix()
```

Match GitHub, maybe make it configurable.